<html>

<head>
<title>Phylogenetic Analysis by Maximum Likelihood (PAML)</title>
</head>

<body link="#0000FF" vlink="#800080" bgcolor="#FFFFFF">
<h2 align="center">Phylogenetic Analysis by Maximum Likelihood (PAML)</h2>

<p align="center"><a href="http://abacus.gene.ucl.ac.uk/"> <b>Ziheng
Yang</b></a> </p>



<hr>

<h2>Table of contents</h2>

<ul>
    <li><a href="#introduction">Introduction </a></li>
    <li><a href="speed.pdf">How fast is my new computer?</a></li>
    <li><a href="#download">Downloading and Compiling PAML</a> </li>
    <li><a href="#running">Running Programs in PAML </a></li>
    <li>To ask questions, go to <a href="http://www.rannala.org/gsf/"><em>Genetics 
    Software Forum</em></a> 
    <li><a href="pamlDOC.pdf">PAML Manual (pamlDOC.pdf)</a>
    <li><a href="pamlFAQs.pdf">PAML FAQs (pamlFAQs.pdf)</a>
    <li><a href="#resources">PAML Resources on the web </a>
    <li><a href="#bugs">Questions and Bug Reports</a> </li>
</ul>


<p>
<h2><a name="introduction">Introduction</a></h2> <p>PAML is a program
package for phylogenetic analyses of DNA or protein sequences using
maximum likelihood. It is maintained and distributed for academic use
free of charge by Ziheng Yang. ANSI C source codes are distributed for
UNIX/Linux/MAC OS X, and executables are provided for MS Windows. </p>

<p>This document is about downloading and compiling PAML and getting
started. See the manual (<a href="pamlDOC.pdf">pamlDOC.pdf</a>) for
more information about running programs in the package.  </p>

<p>Possible uses of the programs are </p>

<ul>
    <li>Estimation of branch lengths in a phylogenetic tree and
        parameters in the evolutionary model such as the
        transition/transversion rate ratio, the shape parameter
        of the gamma distribution for variable evolutionary rates
        among sites, and rate parameters for different genes; </li>
    <li>Test of hypotheses concerning sequence evolution, such as
        rate constancy and independence among nucleotide or amino
        acid sites, rate constancy among lineages (the molecular
        clock), and homogeneity of evolutionary process in
        multiple genes; </li>
    <li>Calculation of substitution rates at sites; </li>
    <li>Reconstruction of ancestral nucleotide or amino acid
        sequences; </li>
    <li>Simulation of nucleotide, codon, and amino acid sequence
        data sets; </li>
    <li>Phylogenetic tree reconstruction by maximum likelihood
        and Bayesian methods (??). </li>
</ul>

<p>A summary of the types of analyses performed by different
programs in the package is given below. </p>

<ul>
    <li><font color="#FF0000" size="4"><code><b>baseml</b></code></font><b>:</b>
        ML analysis of nucleotide sequences: estimation of tree
        topology, branch lengths, and substitution parameters
        under a variety of nucleotide substitution models (JC69,
        K80, F81, F84, HKY85, TN93, REV); constant or gamma rates
        for sites; molecular clock (rate constancy among
        lineages) or no clock, among-gene and within-gene
        variation of substitution rates; models for combined
        analyses of multiple sequence data sets; calculation of
        substitution rates at sites; reconstruction of ancestral
        nucleotides. </li>
    <li><font color="#FF0000" size="4"><code><b>basemlg</b></code></font><b>:</b>
        ML analysis of nucleotide sequences under the model of
        gamma rates among sites. The (continuous) gamma model is
        used with one of the following substitution models: JC69,
        K80, F81, F84, HKY85, TN93, and REV. </li>
    <li><font color="#FF0000" size="4"><code><b>codonml</b></code></font><b>
        (</b><code><b>codeml</b></code><b> with </b><code><b>seqtype
        = 1</b></code><b>):</b> ML analysis of protein-coding DNA
        sequences using codon substitution models (<i>e.g.</i>,
        Goldman and Yang 1994); calculation of the codon-usage
        table; estimation of synonymous and nonsynonymous
        substitution rates; likelihood ratio test of positive
        selection or relaxed selective constraints along lineages
        based on the <i>d</i><sub>N</sub>/<i>d</i><sub>S</sub>
        rate ratios; identification of amino acid sites or
        evolutionary lineages potentially under positive
        selection; reconstruction of ancestral codon sequences. </li>
    <li><font color="#FF0000" size="4"><code><b>aaml</b></code></font><b>
        (</b><code><b>codeml</b></code><b> with </b><code><b>seqtype
        = 2</b></code><b>):</b> ML analysis of amino acid
        sequences under a number of amino acid substitution
        models (Poisson, Proportional, empirical models such as
        those of Dayhoff <i>et al.</i>, Jones <i>et al.</i>,
        mtREV24, and mtmam, and REV); constant or
        gamma-distributed rates among sites; molecular clock
        (rate constancy among lineages) or no clock, among-gene
        and within-gene variation of substitution rates; models
        for combined analyses of multiple gene data; calculation
        of substitution rates at sites; reconstruction of
        ancestral amino acid sequences. </li>
    <li><font color="#FF0000" size="4"><code><b>pamp</b></code></font><b>:</b>
        Parsimony-based analyses for a given tree topology,
        estimation of the substitution pattern by the method of
        Yang and Kumar (1996); estimation of the gamma parameter
        for variable rates among sites by the method of moments,
        the method of Sullivan et al. (1995), and the method of
        Yang and Kumar (1996); reconstruction of ancestral
        character states using the algorithm of Hartigan (1973)
        and an unpublished &quot;improved parsimony&quot; method.
    </li>
    <li><font color="#FF0000" size="4"><code><b>mcmctree</b></code></font><b>:</b>
        Bayesian estimation of phylogenies using DNA sequence
        data (Rannala and Yang, 1996; Yang and Rannala, 1997).
        Markov chain Monte Carlo calculation of posterior
        probabilities of trees.  The algorithm is too slow to be usable.</li>
    <li><font color="#FF0000" size="4"><code><b>evolver</b></code></font><b>:</b>
        This program used to be named <code>listtree</code> and
        does miscellaneous things, such as listing all rooted and
        unrooted trees for a given number of species, generating
        random trees with branch lengths from a birth-death
        process with species sampling, and calculating tree
        bipartition distances. It now also simulates nucleotide,
        codon, or amino acid sequence data sets. Parameters for
        the simulation are specified in the files <code>MCbase.dat</code>,
        <code>MCcodon.dat</code>, and <code>MCaa.dat</code>. You
        can run the program to see the main menu, and then
        consult one of those files to see the details. This
        program can easily fill your hard disk. </li>
    <li><font color="#FF0000" size="4"><b>yn00</b></font>: This program
        implements the method of Yang and Nielsen (2000) for
        estimating synonymous and nonsynonymous substitution
        rates in pairwise comparison of protein-coding DNA
        sequences. The method of Nei and Gojobori (1986) is also
        included in the program. Run yn00 and have a look at the
        control file yn00.ctl and the default result file yn. No
        further documentation is included for this program.</li>
</ul>

<p><b>What does PAML do?</b></p>

<p>PAML is not good for tree making. There are a few options for
heuristic tree search, but they do not work well except for small
data sets of only a few species. If you hope to use PAML to
compare trees from relatively large data sets, one possibility is
to get a collection of candidate trees and then compare them
using more sophisticated models implemented in PAML. You can get
candidate trees by using other programs/methods implemented in
PAUP*, PHYLIP, MOLPHY etc. </p>

<p>PAML may be useful if you are interested in the process of
sequence evolution. The two main programs, baseml and codeml,
implement a number of sophisticated models, which you can use to
construt likelihood ratio tests of evolutionary hypotheses. Right
now, the following options/models do not seem available in other
packages. </p>

<ul>
    <li>Codon-based likelihood analysis for estimating synonymous
        and nonsynonymous rates, and testing hypotheses
        concerning d<sub>N</sub>/d<sub>S</sub> rate ratios
        (Goldman and Yang 1994 MBE 11:725-736; Yang and Nielsen
        1998 JME 46:409-418; Nielsen and Yang 1998 Genetics
        148:929-936; Yang 1998 MBE 15:568-573; Yang et al. 2000
        Genetics 155: 431-449).</li>
    <li>Amino acid-based likelihood analysis with rate variation
        among sites.</li>
    <li>Ancestral sequence reconstruction based on nucleotide,
        codon, or amino acid substitution models (Yang, Kumar,
        and Nei 1995 Genetics 141:1641-1650). Nucleotide sequence
        reconstruction is probably available in DNAML and PAUP*
        too. </li>
    <li>Nucleotide-based models useful for combined analysis of
        multiple gene data (Yang 1996 JME 42:587-596). </li>
    <li>Simulating nucleotide, codon, or amino acid sequence data
        sets. </li>
</ul>

<h2><a name="download">Downloading and Compiling PAML </a></h2>

The downloading ftp site is <a
href="ftp://abacus.gene.ucl.ac.uk/pub/paml/">ftp://abacus.gene.ucl.ac.uk/pub/paml/</a>.
Some network setup does not allow you to visit an anonymous ftp site,
so you may have to speak to your system administrator.

The archive, named something like paml*.*.tar.gz, is for all
platforms.  It includes excutables for MS Windows, and has source
files that you can compile for Mac OSX and unix/linux.


<p><b>Windows 9x/NT/2000/XP. </b> Download the archive.  Unpack it
into a folder, using Winzip, say.  The Windows executables are in the
folder WinEXE/.  The programs are simple Win32 Console applications,
and do not support mice or menu.  Open a "command prompt" box (Start -
Programs - Accessories - Command Prompt).  Run the program by typing
its name rather.  Avoid double-clicking the program names from Windows
Explorer.  You may have to specify the full path, as follows
(depending on where you put the paml folder on your hard disk):<p>

<blockquote>
<code> 
        /paml3.14/winEXE/codeml<br>
	../../paml3.14/winEXE/codeml<br>
</code>
</blockquote>
<p>




<p><b>UNIX, linux, MAC OS X or other systems</b>.  Save the archive on the
  disk.  Unpack it into a folder, as follows<p>

<blockquote>
  <p><code> gzip -d paml3.14.tar.gz </code>
</blockquote>
<p>


Then cd to the paml folder (you have to remember where you saved the
files) and again cd to the src/ folder and compile the programs.<p>

<blockquote><code>
     cd  src <br>
     make -f Makefile <br>
     ls -lF  <br>
     rm *.o  <br>
     mv baseml basemlg codeml pamp evolver yn00 chi2 ..  <br>
     cd ..  <br>
     baseml <br>
     codeml <br>
     evolver <br>
</code></blockquote>

You might have to open and edit the file Makefile before compiling
using make.  For example, you can change cc to gcc and -fast to -O3 or
-O4.  Also see readme.txt in the same folder for compiling
instructions.  You might want to mv the executables into the bin/
folder on your accounts rather than the paml main folder.  And
finally, if your current folder is not on your search path, you will
have to add ./ in front fo the executable file name; that is, use
./codeml instead of codeml to run codeml.

<p><i>MAC OS X</i>. You should open a command terminal
  (Applications-Utilities-Terminal) and then compile and run the
  programs from the terminal.  You cd to the paml folder and then look
  at the readme.txt or Makefile or Makefile.UNIX files.  See above.
  You will need the Mac Developer Toolkit, which is not included in a
  standard installation of OS X and you will get a &quot;Command not
  found&quot; error with either cc or make.  So you should go to the
  Apple web site to download and install the Toolkit
  (http://developer.apple.com/tools/index.html). There are some more
  notes about running programs on MAC OS X or UNIX at the FAQ page.
  <p>For linux administrators, a spec file <a
  href="paml.spec">paml.spec</a> for the linux installer rpm has been
  kindly prepared by Hunter Matthews.


<p><b>PowerMacs (PPC or G3 prior to OS X). </b> Since OS X is now
common, I have stopped distributing executables for MACs running OS 9
or earlier.  MAC executables for two old versions, 3.0a and 3.0c, are
still in the OldVersions/ folder at the ftp site.


<h2><a name="running">Running Programs in PAML </a></h2>

<p>The programs in distribution are essentially the copies I work
on every day, as I make only minor changes before release to the
public. So the programs are not always well tested. Models that I
have never used myself, even it they look sensible or possible
from options in the control file, should be taken with great
caution. I have included example data sets that were used in our
papers for the purpose of error checking. You are encouraged to
duplicate our analysis first to check that the program works and
also to get familiar with the format of the data file and the
interpertation of results.</p>

<p>Programs <font face="Courier New">baseml</font> and <font
face="Courier New">codeml</font> estimate parameters and
calculate the log likelihood values, but do not calcualte the
likelihood ratio statistics. You need to do the subtraction
yourself. The theory is like this. If a more-general model
involves <i>p</i> parameters and has log likelihood <i>l</i><sub>1</sub>,
and a simpler model (which is a special case of the general
model) has <i>q</i> parameters with log-likeliood value <i>l</i><sub>0</sub>,
then 2(<i>l</i><sub>1</sub> - <i>l</i><sub>0</sub>) can be
comared with a chi-square distribution with d.f. = <i>p - q</i>.
Suppose we want to test whether the transition/transversion rate
ratio kappa = 1. We run the JC69 model and get <i>l</i><sub>0</sub>,
and run K80 to get <i>l</i><sub>1</sub>. Then we compare 2(<i>l</i><sub>1</sub>
- <i>l</i><sub>0</sub>) with the chi-square distribution with one
degree of freedom.</p>

<p><b>Running PAML.</b> Most programs in the PAML package have control
files that specify the names of the sequence data file, the tree
structure file, and models and options for the analysis.  The default
control files are <code>baseml.ctl</code> for <code>baseml</code> and
<code>basemlg</code>, <code>codeml.ctl</code> for <code>codeml</code>,
<code>pamp.ctl</code> for <code>pamp</code>, <code>mcmctree.ctl</code>
for <code>mcmctree</code>. The progam <code>evolver</code> does not
have a control file, and uses a simple user interface. All you do is
to type <code>evolver</code> and then choose the options. For other
programs, you should prepare a sequence data file and a tree structure
file, and modify the appropriate control files before running the
programs. The formats of those files are detailed in the documentation
in the package. </p>

<p>You need to prepare a sequence data file (<i>e.g.</i>,
<code>brown.nuc</code>) and modify the options in the appropriate
control file.  If you have chosen <code>runmode</code> = 0 or 1 in the
control file, which means that the tree topologies are specified, you
also need to prepare a <i>tree structure file </i>(<i>e.g.</i>,
<code>trees.4s</code>).  On UNIX or Windows systems, you run the
programs from a command prompt by </p>

<dir>
    <i>ProgramName</i> [for example, baseml]
</dir>

<p>or </p>

<dir>
    <i>ProgramName  ControlFileName </i>
</dir>

<p>On the Mac, you simply click on the program name or icon. You
can do this on a Windows machine too, but it is better if you
open a command box and run the program from there. 



<h1><a name="resources">PAML Resources on the web</a> </h1>
<ul>
    <li><a href="http://www.rannala.org/phpBB2/">PAML discussion group</a>
    <li><a href="http://apt.bea.ki.se">Redhat linux rpm </a>in the Biorpms repository
    <li><a href="pamlResources/readme.txt">A C program for summarizing codeml results for sites under positive selection </a>
</ul>

<h2>Resources for UNIX (Mac OSX) and DOS/Windows </h2>

You need to know a few basics UNIX or DOS commands and know how to run
a program from the command line.  I listed a few in the documentation
included in the package.  There are many little books that teach you
the basics of UNIX, with titles like "UNIX basics", "Teach yourself
UNIX or linux in 24 hours", etc.  Get one of those from the local
bookstore.  There are also a lot of resources on the web.  Search for "UNIX basics", for example.  <p>

<a href="http://workshop.molecularevolution.org/resources/computing.php">Linux
Tutorial from Workshop on Molecular Evolution, Woods Hole</a>


<h2><a name="bugs">Questions and Bug Reports</a></h2>

<p><a href="pamlHistory.txt">Update history and bug fixes </a>
collected here. </p>

<p>Please try to solve your problems by reading the manual
  (pamlDOC.pdf) and the paml FAQ page. Try to identify where the
  problem lies. You should be able to tell from the screen output
  whether the program reads the sequence and tree files correctly and
  to correct mistakes in those files.  

<p>If you can't find answers to your questions in those resources, you
  can post your question at the <em>Genetics Software Forum</em> (<a
  href="http://www.rannala.org/gsf/">http://www.rannala.org/gsf/</a>),
  set up by Dr Bruce Rannala.  This was set up to hopefully reduce the
  amount of time I have to spend in answering user questions.

<p>When reporting problems, please mention the version number, what you
  did and what happened.  In particular copy any error message on the
  screen into the message.  Please use an informative
  &quot;Subject:&quot; field such as "paml/baseml: oom".

<p><b>Try not to send messages to me.  I will try to read and answer
  questions at the discussion site at least once a week.  Messages
  sent to my mail box will be delayed by 2 weeks before I try to
  reply, and I will ignore some messages if the answers are readily
  available from the documentations.  Sorry for the inadequate support
  but I receive more messages than I can answer. </b>

<p>

<p>Some of the nice buttons work. Enjoy clicking on them.</p>
<p>

<table width="100%" border="0" cellpadding="2">
  <tr> 
    <td> 
      <div align="center"><object classid="clsid:D27CDB6E-AE6D-11cf-96B8-444553540000" codebase="http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=4,0,2,0" width="100" height="22" name="introduction">
          <param name=movie value="introductionButton.swf">
          <param name=quality value=high>
          <param name="BGCOLOR" value="">
          <embed src="introductionButton.swf" quality=high pluginspage="http://www.macromedia.com/shockwave/download/index.cgi?P1_Prod_Version=ShockwaveFlash" type="application/x-shockwave-flash" width="100" height="22" bgcolor="" name="Introduction">
          </embed> 
        </object><object classid="clsid:D27CDB6E-AE6D-11cf-96B8-444553540000" codebase="http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=4,0,2,0" width="100" height="22" name="Download">
          <param name=movie value="downloadButton.swf">
          <param name=quality value=high>
          <param name="BGCOLOR" value="">
          <embed src="downloadButton.swf" quality=high pluginspage="http://www.macromedia.com/shockwave/download/index.cgi?P1_Prod_Version=ShockwaveFlash" type="application/x-shockwave-flash" width="100" height="22" bgcolor="" name="Download" base="">
          </embed> 
        </object><object classid="clsid:D27CDB6E-AE6D-11cf-96B8-444553540000" codebase="http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=4,0,2,0" width="100" height="22" name="running">
          <param name=movie value="runningButton.swf">
          <param name=quality value=high>
          <param name="BGCOLOR" value="">
          <embed src="runningButton.swf" quality=high pluginspage="http://www.macromedia.com/shockwave/download/index.cgi?P1_Prod_Version=ShockwaveFlash" type="application/x-shockwave-flash" width="100" height="22" bgcolor="" name="running">
          </embed> 
        </object><object classid="clsid:D27CDB6E-AE6D-11cf-96B8-444553540000" codebase="http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=4,0,2,0" width="100" height="22" name="Manual">
          <param name="BASE" value=".">
          <param name=movie value="manualButton.swf">
          <param name=quality value=high>
          <param name="BGCOLOR" value="#FF0000">
          <embed src="manualButton.swf" quality=high pluginspage="http://www.macromedia.com/shockwave/download/index.cgi?P1_Prod_Version=ShockwaveFlash" type="application/x-shockwave-flash" width="100" height="22" bgcolor="#FF0000" name="Manual" base=".">
          </embed> 
        </object><object classid="clsid:D27CDB6E-AE6D-11cf-96B8-444553540000" codebase="http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=4,0,2,0" width="100" height="22" name="FAQs">
          <param name="BASE" value=".">
          <param name=movie value="faqsButton.swf">
          <param name=quality value=high>
          <param name="BGCOLOR" value="">
          <embed src="faqsButton.swf" quality=high pluginspage="http://www.macromedia.com/shockwave/download/index.cgi?P1_Prod_Version=ShockwaveFlash" type="application/x-shockwave-flash" width="100" height="22" bgcolor="" name="FAQs" base=".">
          </embed> 
        </object><object classid="clsid:D27CDB6E-AE6D-11cf-96B8-444553540000" codebase="http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=4,0,2,0" width="100" height="22">
          <param name=movie value="bugButton.swf">
          <param name=quality value=high>
          <param name="BGCOLOR" value="">
          <embed src="bugButton.swf" quality=high pluginspage="http://www.macromedia.com/shockwave/download/index.cgi?P1_Prod_Version=ShockwaveFlash" type="application/x-shockwave-flash" width="100" height="22" bgcolor="">
          </embed> 
        </object></div>
    </td>
  </tr>
</table>
<p><object classid="clsid:D27CDB6E-AE6D-11cf-96B8-444553540000" codebase="http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=4,0,2,0" width="280" height="35">
    <param name=movie value="speedButton.swf">
    <param name=quality value=high>
    <param name="BASE" value=".">
    <param name="BGCOLOR" value="">
    <param name="SCALE" value="exactfit">
    <embed src="speedButton.swf" base="."  quality=high pluginspage="http://www.macromedia.com/shockwave/download/index.cgi?P1_Prod_Version=ShockwaveFlash" type="application/x-shockwave-flash" scale="exactfit" width="280" height="35" bgcolor="">
    </embed> 
  </object></p>


<p>Counter since 8 March 2002. 
  <!--Begin Ultimate Counter code-->
</p>
<p></p>
<center>
<a target="_top" href="http://www.UltimateCounter.com/link.cgi?id=495833">
<img src = "http://www.UltimateCounter.com/count/count.cgi?495833">
</a></center><!--End Ultimate Counter code--> 

</body>
</html>
